### PR TITLE
[PF-2155] Allow admins to modify users' profiles

### PR DESCRIPTION
**Pull request by iliadsh**

### CODE REVIEW

File: service/src/main/resources/api/openapi.yml [hunk 1]
```diff
@@ -132,6 +137,13 @@ components:
       schema:
         $ref: '#/components/schemas/LookupPath'
 
+    UserEmail:
+      name: userEmail
+      in: query
+      description: Target a specific user by their email address.
```
Comment (yuhuyoyo) [inline service/src/main/resources/api/openapi.yml:146]: required false?

File: service/src/main/resources/api/openapi.yml [hunk 2]
```diff
@@ -62,6 +65,8 @@ paths:
 
         ProfilePath must not be empty, as it is an error to attempt to overwrite the entire
         profile object. On success, the value to set will be returned.
+
+        Supplying an email will target that user's profile (requires Admin).
```
Comment (yuhuyoyo) [inline service/src/main/resources/api/openapi.yml]: If left blank, will target the current user.

File: service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java [hunk 3]
```diff
@@ -19,32 +21,44 @@ public class ProfileApiController implements ProfileApi {
...
     this.servletRequest = servletRequest;
     this.profileService = profileService;
     this.userFactory = userFactory;
+    this.samService = samService;
   }
 
   @Override
-  public ResponseEntity<AnyObject> setUserProfile(AnyObject body, String path) {
-    profileService.setProperty(getUser(), parsePath(path), body.getValue());
+  public ResponseEntity<AnyObject> setUserProfile(AnyObject body, String path, String userEmail) {
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java]: annotate nullable

File: service/src/main/java/bio/terra/user/service/iam/SamService.java [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+
+  public UsersApi samUsersApi(String accessToken) {
+    return new UsersApi(getApiClient(accessToken));
+  }
+
+  public AdminApi samAdminApi(String accessToken) {
+    return new AdminApi(getApiClient(accessToken));
+  }
+
+  public String adminEmailToId(BearerToken userRequest, String email) throws InterruptedException {
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/user/service/iam/SamService.java]: nit: adminGetUserIdByEmail

File: service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java [hunk 5]
```diff
@@ -19,32 +21,44 @@ public class ProfileApiController implements ProfileApi {
...
 
     return new ResponseEntity<>(apiObj, HttpStatus.OK);
   }
 
+  private String targetIdOrSelf(String userEmail) {
+    return StringUtils.isEmpty(userEmail)
+        ? getUser().getSubjectId()
+        : SamRethrow.onInterrupted(
+            () -> samService.adminEmailToId(getUser().getBearerToken(), userEmail),
+            "Admin user Id from email");
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java]: rephrase it to something like If user is admin, get the targeted user id from email.

It's easy to misread this as getting the admin's user id.

File: service/src/main/java/bio/terra/user/service/iam/SamRethrow.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
+package bio.terra.user.service.iam;
+
+import bio.terra.common.sam.exception.SamExceptionFactory;
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/user/service/iam/SamRethrow.java:3]: you don't *have* to do this but there are at least three copies of SamRethrow in Databiosphere now.. https://cs.github.com/?scopeName=All+repos&scope=&q=samrethrow

it should live with SamRetry here https://cs.github.com/DataBiosphere/terra-common-lib/blob/84d0db138bd713a339391519686308e160b58187/src/main/java/bio/terra/common/sam/SamRetry.java?q=samretry

File: service/src/main/java/bio/terra/user/service/iam/SamService.java [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+    ApiClient apiClient = getApiClient();
+    apiClient.setAccessToken(accessToken);
+    return apiClient;
+  }
+
+  private ApiClient getApiClient() {
+    return new ApiClient().setBasePath(samConfig.basePath());
+  }
+
+  public UsersApi samUsersApi(String accessToken) {
```
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/user/service/iam/SamService.java:35]: not used?

File: service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java (hunk 3 above)
Comment (yuhuyoyo) [inline service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java]: not strictly related to this PR, but i think we should always check if the user is a sam user for all the endpoint?

File: service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java [hunk 8]
```diff
@@ -26,6 +28,8 @@ class ProfileApiControllerTest extends BaseUnitTest {
 
   @MockBean private ProfileSamUserFactory userFactory;
 
+  @MockBean private SamService samService;
+
   private final SamUser user = mock(SamUser.class);
```
Comment (yuhuyoyo) [inline service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java]: nit: you can also use @Mock for this to be consistent with the style.

**Review by yuhuyoyo — APPROVED** [review]

File: service/src/main/resources/api/openapi.yml (hunk 2 above)
Comment (melissachang) [reply service/src/main/resources/api/openapi.yml]: If userEmail is set, requester must be SAM admin, and userEmail's profile is returned. If userEmail is not set, requester's profile is returned.

File: service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java (hunk 3 above)
Comment (melissachang) [inline service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java]: We have a convention that the first thing controller does, is check access control. [Example.](https://cs.github.com/DataBiosphere/terra-workspace-manager/blob/main/service/src/main/java/bio/terra/workspace/app/controller/WorkspaceApiController.java#L298)

[Here's an example of checking if requester is SAM admin.](https://github.com/DataBiosphere/terra-workspace-manager/blob/main/service/src/main/java/bio/terra/workspace/app/controller/AdminApiController.java#L36-L39)

So you would have:

```
if (StringUtils.isEmpty(userEmail)) {
      AuthenticatedUserRequest userRequest = getAuthenticatedInfo();
    SamRethrow.onInterrupted(
        () -> getSamService().checkAdminAuthz(userRequest),
        "check whether the user has admin access");
}
```

File: service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java [hunk 9]
```diff
@@ -19,32 +21,44 @@ public class ProfileApiController implements ProfileApi {
...
     this.profileService = profileService;
     this.userFactory = userFactory;
+    this.samService = samService;
   }
 
   @Override
-  public ResponseEntity<AnyObject> setUserProfile(AnyObject body, String path) {
-    profileService.setProperty(getUser(), parsePath(path), body.getValue());
+  public ResponseEntity<AnyObject> setUserProfile(AnyObject body, String path, String userEmail) {
+    profileService.setProperty(targetIdOrSelf(userEmail), parsePath(path), body.getValue());
```
Comment (melissachang) [inline service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java]: targetIdOrSelf -> getProfileUser ?

Also I think it's a tad more readable to not inline:

```
String profileUser = getProfileUser(userEmail);
profileService.setProperty(profileUser, parsePath(path), body.getValue());
```

File: service/src/main/java/bio/terra/user/service/iam/SamService.java (hunk 7 above)
Comment (iliadsh) [reply service/src/main/java/bio/terra/user/service/iam/SamService.java:35]: Thanks, forgot to clean it up after testing

File: service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java (hunk 3 above)
Comment (iliadsh) [reply service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java]: Agree, though I think this is already implicitly happening with the current endpoints since they either ask for the current SamUser or attempt to make an admin request where they ofc. have to be a user

File: service/src/main/java/bio/terra/user/service/iam/SamRethrow.java (hunk 6 above)
Comment (iliadsh) [reply service/src/main/java/bio/terra/user/service/iam/SamRethrow.java:3]: Just to confirm, you're suggesting contributing SamRethrow to the common lib right? If so I agree and was surprised to find it not there already

File: service/src/main/java/bio/terra/user/service/iam/SamService.java [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+      throws InterruptedException {
+    try {
+      return SamRetry.retry(
+          () ->
+              samAdminApi(userRequest.getToken())
+                  .adminGetUserByEmail(email)
+                  .getUserInfo()
+                  .getUserSubjectId());
+    } catch (ApiException apiException) {
+      throw new ForbiddenException("Terra Admin access is required for this action.");
```
Comment (melissachang) [inline service/src/main/java/bio/terra/user/service/iam/SamService.java]: How about

Must be SAM admin to read or write another user's profile

It's a little more precise... admin access it not required if userEmail isn't set.

File: service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java [hunk 11]
```diff
@@ -71,10 +75,44 @@ void multiUser() throws Exception {
     assertUserProfile("$.value.name", "John");
   }
 
+  @Test
+  void adminNoPermission() throws Exception {
+    when(samService.adminGetUserIdByEmail(any(), any())).thenThrow(new ForbiddenException(""));
+    mockMvc
```
Comment (melissachang) [inline service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java]: nit Write helper method. It makes test easier to read, and another test could use the method in the future.

```
  @Test
  void adminNoPermission() throws Exception {
      when(samService.adminGetUserIdByEmail(any(), any())).thenThrow(new ForbiddenException(""));
      setUserProfileExpectCode("name", "{ \"value\": \"John\" }", "user2@gmail.com", HttpStatus.SC_FORBIDDEN);
  }
```

File: service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java (hunk 3 above)
Comment (yuhuyoyo) [reply service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java]: ah you are right

File: service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java [hunk 12]
```diff
@@ -71,10 +75,44 @@ void multiUser() throws Exception {
     assertUserProfile("$.value.name", "John");
   }
 
+  @Test
```
Comment (melissachang) [inline service/src/test/java/bio/terra/user/app/controller/ProfileApiControllerTest.java:79]: opt nit tests are easier to read with this style naming. If you're testing one method, test name starts with the method that you're testing.

```
getUserProfile_emptyProfile
setUserProfile_requestEmailNotSet
getUserProfile_noValue
multiUser
getUserProfile_requestEmailSet_requesterNotAdmin_throws403
requestEmailSet_requesterIsAdmin


```

**Review by melissachang — APPROVED** [review]

File: service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java (hunk 3 above)
Comment (iliadsh) [reply service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java]: Fair enough, though that's pretty much the purpose of `getProfileUser`, just merged into one sam call rather than two by checking auth and getting the user id at once

File: service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java [hunk 13]
```diff
@@ -19,32 +22,47 @@ public class ProfileApiController implements ProfileApi {
...
     this.servletRequest = servletRequest;
     this.profileService = profileService;
     this.userFactory = userFactory;
+    this.samService = samService;
   }
 
   @Override
-  public ResponseEntity<AnyObject> setUserProfile(AnyObject body, String path) {
-    profileService.setProperty(getUser(), parsePath(path), body.getValue());
+  public ResponseEntity<AnyObject> setUserProfile(
```
Comment (melissachang) [inline service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java:40]: I still think the code should be more clear that it's checking auth in the beginning. If you look at WSM WorkspaceApiController.java, every method either checks auth in the beginning (validateWorkspaceAndAction()), or has a comment explaining why auth is not checked. This way, we know where to look to confirm that we're checking auth; we look in controller, not service or dao. I think it's ok to make to make an extra SAM call for this.

Creating SamService.checkAdminAuthz is optional -- somehow the code in setUserProfile(), getUserProfile() should be clear that it's checking auth. Some options.

```
if (!StringUtils.isEmpty(userEmail)) {
      AuthenticatedUserRequest userRequest = getAuthenticatedInfo();
    SamRethrow.onInterrupted(
        () -> getSamService().checkAdminAuthz(userRequest),
        "check whether the user has admin access");
}
```

Or put above code into ProfileApiController.maybeCheckRequesterIsSamAdmin() and do:

```
maybeCheckRequesterIsSamAdmin();
```

Or create ProfileApiController.checkRequesterIsSamAdmin() and:

```
if (!StringUtils.isEmpty(userEmail)) {
  checkRequesterIsSamAdmin();
}
```

File: service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java [hunk 14]
```diff
@@ -19,32 +22,47 @@ public class ProfileApiController implements ProfileApi {
...
+      @Nullable String path, @Nullable String userEmail) {
+    String profileUser = getProfileUser(userEmail);
     AnyObject apiObj =
-        new AnyObject().value(profileService.getProperty(getUser(), parsePath(path)));
+        new AnyObject().value(profileService.getProperty(profileUser, parsePath(path)));
 
     return new ResponseEntity<>(apiObj, HttpStatus.OK);
   }
 
+  private String getProfileUser(String userEmail) {
```
Comment (melissachang) [inline service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java:65]: /** Gets user email to perform profile action on. */

**Review by melissachang — APPROVED** [review]

File: service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java (hunk 13 above)
Comment (iliadsh) [reply service/src/main/java/bio/terra/user/app/controller/ProfileApiController.java:40]: Thanks! I added the `checkRequesterIsSamAdmin` check.
